llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp

```diff
@@ -204,10 +204,21 @@ void RISCVRegisterInfo::adjustReg(MachineBasicBlock &MBB,
     uint32_t NumOfVReg = ScalableValue / 8;
     BuildMI(MBB, II, DL, TII->get(RISCV::PseudoReadVLENB), ScratchReg)
         .setMIFlag(Flag);
-    TII->mulImm(MF, MBB, II, DL, ScratchReg, NumOfVReg, Flag);
-    BuildMI(MBB, II, DL, TII->get(ScalableAdjOpc), DestReg)
-      .addReg(SrcReg).addReg(ScratchReg, RegState::Kill)
-      .setMIFlag(Flag);
+
+    if (ScalableAdjOpc == RISCV::ADD && ST.hasStdExtZba() &&
+        (NumOfVReg == 2 || NumOfVReg == 4 || NumOfVReg == 8)) {
+      unsigned Opc = NumOfVReg == 2 ? RISCV::SH1ADD :
+        (NumOfVReg == 4 ? RISCV::SH2ADD : RISCV::SH3ADD);
+      BuildMI(MBB, II, DL, TII->get(Opc), DestReg)
+          .addReg(ScratchReg, RegState::Kill)
+          .addReg(SrcReg, getKillRegState(KillSrcReg))
```

Comment:
If we're adding `getKillRegState(KillSrcReg)` here, should we do it below in the `else` too, for consistency? It always looks `false` so maybe that's why it wasn't done before.

Reply:
We do have it on the else case. It's on the add at line 219.

Reply:
I had to do a double-take, but I meant on `SrcReg`.

Reply:
Oops, you're right.  I misread.  Will follow up with a separate commit.

Reply:
Addressed in 5ae9ffbd18fd93edbbc8efebe140aeb24cd763c2.  Sorry for the confusion here, I misread my own code when looking at your comment.  I think it was a combination of the swapped operand order, and that the shXadd case below does use the KillSrcReg flag.  But still, sorry for the unhelpful and confusing initial response.

Reply:
No worries, thanks for the follow-up.